LF: Fix specification of ArithmeticError (#9949)
digital-asset/daml: daml-lf/spec/daml-lf-1.rst

The functions come with their types and a description of their
behavior.

About Exceptions
~~~~~~~~~~~~~~~~

Some builtin functions can throw non-fatal exceptions, i.e. exceptions
catchable by the ``TryCatch`` update expression. Those exceptions are
not built in the language but are standard exceptions defined in user
land. The builtin functions from an engine compliant with the current
specification should be able to produce and handle (notably the

> ``ANY_EXCEPTION_MESSAGE`` builtin function) such exceptions without
> the packages where those exception are defined to be explicitly
> loaded.  Any other usage like on the exception payload, like

Reviewer: ```suggestion
``ANY_EXCEPTION_MESSAGE`` builtin function) such exceptions even if the package they are defined in has not been loaded.
```

construction, projection, update or conversion from/back
`'AnyException'`, requires the definition packages to be loaded.

As of LF 1.14 the only non-fatal exceptions that a builtin function
can throw is the ``ArithmeticError`` record defined in the module
``DA.Exception.ArithmeticError`` of the package
``'cb0552debf219cc909f51cbb5c3b41e9981d39f8f645b1f35e2ef5be2e0b858a'``
whose content is as follow::

   package cb0552debf219cc909f51cbb5c3b41e9981d39f8f645b1f35e2ef5be2e0b858a
   daml-lf 1.14
   metadata daml-prim-DA-Exception-ArithmeticError-1.0.0

   module DA.Exception.ArithmeticError {
      record @serializable ArithmeticError ↦ { message : Text } ;
      val $WArithmeticError : Text -> DA.Exception.ArithmeticError:ArithmeticError =
         λ message : Text .
            DA.Exception.ArithmeticError:ArithmeticError { message = message };
      exception ArithmeticError ↦ {
         'message' λ x : DA.Exception.ArithmeticError:ArithmeticError.
            DA.Exception.ArithmeticError:ArithmeticError { message } x
      } ;
   }

The package can be produced in a stable way by Daml SDK 1.14 or latter
with the command::

>   bazel build //compiler/damlc/stable-packages:stable-packages

Reviewer: not sure that makes sense in a spec.

    unless both neighbors are equidistant, in which case round towards
    the even neighbor.

>   - ``'ROUNDING_UNNECESSARY'`` : Throws ``ArithmeticError`` if the

Reviewer: that change seems off, `ROUNDING_UNNECSSARY` does not throw. It causes the bulitin you pass it in to throw.

Author: I am not sure what is the issue here. 
I just added
- a "s" at the end of "Throw" 
- two "`" 
- and a newline

Reviewer: With the `s` it sounds  to me like `ROUNDING_UNNECESSARY` is a primitive that throws rather than an argument which causes something else to throw. But that’s mostly gut feel so maybe my language intuition is just off here.

Author: OK. I revert back and remove the 's'. I also removing from some of the "Round" there were in the enumeration.
